cirq/circuits/circuit_operation.py

```diff
@@ -188,53 +269,95 @@ def _json_dict_(self):
         )
 
     # Methods for constructing a similar object with one field modified.
 
     def repeat(
         self,
-        repetitions: INT_TYPE,
+        repetitions: Optional[INT_TYPE] = None,
+        repetition_ids: Optional[List[str]] = None,
     ) -> 'CircuitOperation':
         """Returns a copy of this operation repeated 'repetitions' times.
+         Each repetition instance will be identified by a single repetition_id.
 
         Args:
             repetitions: Number of times this operation should repeat. This
-                is multiplied with any pre-existing repetitions.
+                is multiplied with any pre-existing repetitions. If unset, it
+                defaults to the length of `repetition_ids`.
+            repetition_ids: List of IDs, one for each repetition. If unset,
+                defaults to `default_repetition_ids(repetitions)`.
 
         Returns:
-            A copy of this operation repeated 'repetitions' times.
+            A copy of this operation repeated `repetitions` times with the
+            appropriate `repetition_ids`. The output `repetition_ids` are the
+            cartesian product of input `repetition_ids` with the base
+            operation's `repetition_ids`. If the base operation has unset
+            `repetition_ids` (indicates {-1, 0, 1} `repetitions` with no custom
+            IDs), the input `repetition_ids` are directly used.
 
         Raises:
             TypeError: `repetitions` is not an integer value.
-            NotImplementedError: The operation contains measurements and
-                cannot have repetitions.
+            ValueError: Unexpected length of `repetition_ids`.
+            ValueError: Both `repetitions` and `repetition_ids` are None.
         """
+        if repetitions is None:
+            if repetition_ids is None:
+                raise ValueError('At least one of repetitions and repetition_ids must be set')
+            repetitions = len(repetition_ids)
+
         if not isinstance(repetitions, (int, np.integer)):
             raise TypeError('Only integer repetitions are allowed.')
-        if repetitions == 1:
+
+        repetitions = int(repetitions)
+
+        if repetitions == 1 and repetition_ids is None:
             # As CircuitOperation is immutable, this can safely return the original.
             return self
-        repetitions = int(repetitions)
-        if protocols.is_measurement(self.circuit):
-            raise NotImplementedError('Loops over measurements are not supported.')
-        return self.replace(repetitions=self.repetitions * repetitions)
+
+        expected_repetition_id_length = abs(repetitions)
+        # The eventual number of repetitions of the returned CircuitOperation.
+        final_repetitions = self.repetitions * repetitions
+
+        if repetition_ids is None:
+            repetition_ids = default_repetition_ids(repetitions)
+        elif len(repetition_ids) != abs(expected_repetition_id_length):
```

Comment:
`abs` not needed here? (already called when initializing `expected_repetition_id_length`)

Reply:
Done